sklearn/neighbors/dist_metrics.pyx

```diff
@@ -1111,7 +1100,13 @@ cdef class PyFuncDistance(DistanceMetric):
         with gil:
             x1arr = _buffer_to_ndarray(x1, size)
             x2arr = _buffer_to_ndarray(x2, size)
-            return self.func(x1arr, x2arr, **self.kwargs)
+            d = self.func(x1arr, x2arr, **self.kwargs)
+            try:
+                return d
+            except TypeError:
```

Comment:
Maybe it's my knowledge of Python that is lacking, but I am unsure that the exception will be really captured in the way I think.

Indeed, if I understand things correctly, the exception that we are expecting will be raised outside this function. Hence, I suspect that the try/except will not trigger. It seems to me that, whether it triggers or not is depends on the semantics of in which frame the exception is raised.

I think that it would be great to have a test that shows us that the exception is indeed raised, given that it is not trivial.